Index: lib/SimpleSAML/Metadata/MetaDataStorageHandlerFlatFile.php
===================================================================
--- lib/SimpleSAML/Metadata/MetaDataStorageHandlerFlatFile.php	(revision 3231)
+++ lib/SimpleSAML/Metadata/MetaDataStorageHandlerFlatFile.php	(working copy)
@@ -113,30 +113,8 @@
 
 		return $metadataSet;
 	}
-	
-	private function generateDynamicHostedEntityID($set) {
 
-		/* Get the configuration. */
-		$baseurl = SimpleSAML_Utilities::getBaseURL();
 
-		if ($set === 'saml20-idp-hosted') {
-			return $baseurl . 'saml2/idp/metadata.php';
-		} elseif($set === 'saml20-sp-hosted') {
-			return $baseurl . 'saml2/sp/metadata.php';			
-		} elseif($set === 'shib13-idp-hosted') {
-			return $baseurl . 'shib13/idp/metadata.php';
-		} elseif($set === 'shib13-sp-hosted') {
-			return $baseurl . 'shib13/sp/metadata.php';
-		} elseif($set === 'wsfed-sp-hosted') {
-			return 'urn:federation:' . SimpleSAML_Utilities::getSelfHost();
-		} elseif($set === 'adfs-idp-hosted') {
-			return 'urn:federation:' . SimpleSAML_Utilities::getSelfHost() . ':idp';
-		} else {
-			throw new Exception('Can not generate dynamic EntityID for metadata of this type: [' . $set . ']');
-		}
-	}
-
-
 }
 
 ?>
\ No newline at end of file
Index: lib/SimpleSAML/Metadata/MetaDataStorageSource.php
===================================================================
--- lib/SimpleSAML/Metadata/MetaDataStorageSource.php	(revision 3231)
+++ lib/SimpleSAML/Metadata/MetaDataStorageSource.php	(working copy)
@@ -224,6 +224,28 @@
 
 		return NULL;
 	}
+	
+	protected function generateDynamicHostedEntityID($set) {
 
+		/* Get the configuration. */
+		$baseurl = SimpleSAML_Utilities::getBaseURL();
+
+		if ($set === 'saml20-idp-hosted') {
+			return $baseurl . 'saml2/idp/metadata.php';
+		} elseif($set === 'saml20-sp-hosted') {
+			return $baseurl . 'saml2/sp/metadata.php';			
+		} elseif($set === 'shib13-idp-hosted') {
+			return $baseurl . 'shib13/idp/metadata.php';
+		} elseif($set === 'shib13-sp-hosted') {
+			return $baseurl . 'shib13/sp/metadata.php';
+		} elseif($set === 'wsfed-sp-hosted') {
+			return 'urn:federation:' . SimpleSAML_Utilities::getSelfHost();
+		} elseif($set === 'adfs-idp-hosted') {
+			return 'urn:federation:' . SimpleSAML_Utilities::getSelfHost() . ':idp';
+		} else {
+			throw new Exception('Can not generate dynamic EntityID for metadata of this type: [' . $set . ']');
+		}
+	}
+
 }
 ?>
\ No newline at end of file
